### ISSUE TITLE
Add 'tip diameter' to nozzle tip wizard

### ISSUE THREAD

**Issue by dzach** [openpnp/openpnp#580]
# Feature Request
## Describe the Feature
Add an entry for setting the tip diameter for each nozzle and have it available to the pipeline and the rest of the system. It can be used in many places to represent the size of the tip nozzle, e.g. as a reticle, as a center mark in the pipeline image, as the COG in models, e.g. rotated rects, helping diagnose incompatibility of nozzle tip with part size:

![screenshot](https://user-images.githubusercontent.com/1109829/26934719-da4ea336-4c72-11e7-98c0-873cf4e97f99.png)

## How Will It Look?
If the feature will include changes to the user interface, how would they look? Include drawings or mockups if possible.

![tipdiameter](https://user-images.githubusercontent.com/1109829/26934657-b2030502-4c72-11e7-99d1-1b732600353c.png)

## Why Do We Need It?
1. What problem does it solve?
Today, we are not able to directly compare nozzle tip size with part size. The proposed feature will provide visual clues for such comparison.
2. Is it useful for everyone, or does it only solve a problem for your machine?
This is for the benefit of everyone using OpenPnP.

**Comment by vonnieda** [comment]
Great idea!

**Comment by dzach** [comment]
Ok, I've got my `infinite monkey coder (TM)`to go as far as getting the label and the text box to appear in the wizard. 

![screenshot-1](https://user-images.githubusercontent.com/1109829/26940139-43fbdfe2-4c82-11e7-9aa8-b1e69fb5f81e.png)

There is a chance I have also succeeded to properly include the new property `tipDiameter` to the `AbstractNozzleTip` class.

How do I get the value of the tip diameter, say from a pipeline script stage? `pipeline.getNozzle()` returns just `null`. Is this correct?

I can reach my `tipDiameter` with this:
```
pipeline.getCamera().getHead().getDefaultNozzle().getNozzleTip().getTipDiameter()
```
but this doesn't feel right. Or does it?

**Comment by vonnieda** [comment]
Nozzle has to be set on the pipeline before it's processed. So, in whatever class you want to use this, you need to call pipeline.setNozzle(the nozzle) before calling pipeline.process().

**Comment by dzach** [comment]
Yes, now that I think of it, it is logical. In my example with the pipeline: a pipeline does not have a direct connection to a nozzle, but it does have one with the camera that is grabbing the image it is processing. So, I guess, it does feel right to get the tipDiameter through the camera, after all.

Should I submit this little addition of the nozzle tip diameter with a PR?

**Comment by vonnieda** [comment]
Sorry if I wasn't clear. CvPipeline does have a nozzle property (getNozzle, setNozzle) but the nozzle has to be set in whatever is calling the pipeline. For instance, in LoosePartFeeder, this happens around line 83. If the thing that is calling the pipeline doesn't set the nozzle, then the pipeline will not have access to a nozzle. It is not a good idea to use getDefaultNozzle since that will likely be wrong on a system with multiple nozzles.

For example, let's say your system has 2 nozzles. The system goes to pick the first nozzle, and you call getDefaultNozzle(). That works fine. But then for the second nozzle you call getDefaultNozzle and you end up with the data for the first nozzle.

So, in short, if you need access to the nozzle in the pipeline, you should use pipeline.getNozzle(), and for any pipeline that uses that it must call setNozzle beforehand.

**Comment by vonnieda** [comment]
Oh, and yes, please submit a PR for the nozzle tip diameter change. It will need to be a Length value, and not a double, though. So make sure that's the case first. A double has no units associated with it, a Length does.

**Comment by dzach** [comment]
> pipeline.getNozzle() returns just null. Is this correct?

I've already checked that but it returns `null`, while the pipeline is called by `Edit pipeline` of the LoosePartFeeder wizard.

**Comment by vonnieda** [comment]
Yea, looks like those properties are not being set properly in the wizard.
They were added to the feeder, but not the wizard. I'd suggest adding

        pipeline
.setNozzle(MainFrame.get().getMachineControls().getSelectedNozzle());


at line 96 in ReferenceLoosePartFeederConfigurationWizard.

Jason

**Comment by dzach** [comment]
I almost have it, but I don't want to fiddle with the tests:
```
testSampleJob(SampleJobTest): Error while saving machine.xml (Transform of class org.openpnp.model.Length not supported)
```

**Comment by dzach** [comment]
It comes from here:
``` 
   public AbstractNozzleTip() {
        this.id = Configuration.createId("TIP");
        this.name = getClass().getSimpleName();
        this.tipDiameter = new Length(1.0, Configuration.get().getSystemUnits());
    }
```

**Comment by vonnieda** [comment]
Can you push the code to a branch so I can look at it? That looks like it
should work fine.

**Comment by vonnieda** [comment]
Oh, make sure you have it set as an @Element and not an @Attribute.

Jason

**Comment by dzach** [comment]
> Oh, make sure you have it set as an `@Element`

That did it.

https://github.com/dzach/openpnp/tree/feature-NozzleTipDiameter
